### PR TITLE
Structured concurrency

### PR DESCRIPTION
**Pull request by c42f**
A document describing structured concurrency and (eventually) how it might be applied in julia.

**Please direct substantive comments to the JuliaLang/julia issue https://github.com/JuliaLang/julia/issues/33248**

(procedural errors such as spelling may be pointed out here)

### CODE REVIEW

**Review by StefanKarpinski — APPROVED** [review]

File: StructuredConcurrency.md [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+  concurrency thanks to `@njsmith`'s tutorial at PyCon 2018.
+  - [General discussion](https://trio.discourse.group/t/happy-eyeballs-structured-concurrencys-hello-world/57)
+  - [Trio implementation](https://github.com/python-trio/trio/blob/master/trio/_highlevel_open_tcp_stream.py)
+  - [Libdill implementation](https://github.com/sustrik/libdill/blob/master/happyeyeballs.c) and [discussion](http://250bpm.com/blog:139)
+* The Go concurrency tutorial — in his talk `@elizarov` suggested that
+  implementing all the examples there was a great inspiriation.
+
+## Related julia issues and prototypes
+
+* [Tapir parrallel IR](https://github.com/JuliaLang/julia/pull/31086)
```
Comment (vchuravy) [inline StructuredConcurrency.md]: Note that this is primarily about:
- Compiler representation of tasks
- Experimenting with Cilk-style semantics (e.g. may happen in parallel or what I have taken to call structured parallelism)

Comment (c42f) [reply StructuredConcurrency.md]: Thanks, I've replied to this in the main repo.

**Comment by c42f** [comment]
I have merged this; not because I think it's complete in any way, but to give us a base to work on.

File: StructuredConcurrency.md [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+leaves us in a difficult situation: how do we allow for cancellation of
+expensive numerical operations? Are there options for cancellation of numerical
+loops with a semantic which can be understood by users? The Go people seem to
+consider that arbitrary *preemption* is workable, but can arbitrary
+cancellation be made to work with the right language and library features?
+
+#### Runtime technicalities for preemption
+
+On a technical level, our runtime situation in julia-1.3 is very similar to
+Go where preemption is cooperative and a rouge goroutine can sometimes wedge
```
Comment (ArchRobison) [inline StructuredConcurrency.md:173]: Typo: `rouge` -> `rogue`

Comment (c42f) [reply StructuredConcurrency.md:173]: Hi Arch, thanks; will be fixed in the next revision. If you do have time to provide some perspective, please head on over to

https://github.com/JuliaLang/julia/issues/33248

where I am pulling on a lot of disparate threads and undoubtedly need all the help I can get from people who actually know things :-)

**Pull request merged**